pydra/utils/typing.py

```diff
@@ -0,0 +1,603 @@
+
+
+T = ty.TypeVar("T")
+TypeOrAny = ty.Union[type, ty.Any]
+
+
+class TypeParser(ty.Generic[T]):
+    """A callable which can be used as a converter for attrs.fields to check whether an
+    object or LazyField matches the specified field type, or can be
+    coerced into it (given the criteria passed on initialisation of the checker).
+    Nested container type are expanded and each of their type args are checked/coerced
+    against corresponding parts of the object.
+
+    Parameters
+    ----------
+    tp : type
+        the type objects will be coerced to
+    coercible: Iterable[ty.Tuple[type or Any, type or Any]], optional
+        limits coercing between the pairs of types where they appear within the
+        tree of more complex nested container types. If None, then all types are
+        coercible except explicitly excluded
+    not_coercible: Iterable[ty.Tuple[type or Any, type or Any]], optional
+        excludes the limits coercing between the pairs of types where they appear within
+        the tree of more complex nested container types. Overrides 'coercible' to enable
+        you to carve out exceptions, such as TypeParser(list, coercible=[(ty.Iterable, list)],
+        not_coercible=[(str, list)])
+    """
+
+    tp: ty.Type[T]
+    coercible: ty.List[ty.Tuple[TypeOrAny, TypeOrAny]]
+    not_coercible: ty.List[ty.Tuple[TypeOrAny, TypeOrAny]]
+
+    COERCIBLE_DEFAULT = (
+        (ty.Sequence, ty.Sequence),
+        (ty.Mapping, ty.Mapping),
+        (Path, os.PathLike),
+        (str, os.PathLike),
+        (os.PathLike, Path),
+        (os.PathLike, str),
+        (ty.Any, MultiInputObj),
+        (ty.Union[os.PathLike, str], MultiInputFile),
+        (ty.Sequence, MultiOutputObj),
+        (int, float),
+    )
+
+    NOT_COERCIBLE_DEFAULT = ((str, ty.Sequence), (ty.Sequence, str))
+
+    def __init__(
+        self,
+        tp,
+        coercible: ty.Optional[
+            ty.Iterable[ty.Tuple[TypeOrAny, TypeOrAny]]
+        ] = COERCIBLE_DEFAULT,
+        not_coercible: ty.Optional[
+            ty.Iterable[ty.Tuple[TypeOrAny, TypeOrAny]]
+        ] = NOT_COERCIBLE_DEFAULT,
+    ):
+        def expand_pattern(t):
+            """Recursively expand the type arguments of the target type in nested tuples"""
+            if t is inspect._empty:
+                return None
+            origin = get_origin(t)
+            if origin is None:
+                return t
+            args = get_args(t)
+            if not args or args == (Ellipsis,):
+                assert isinstance(origin, type)
+                return origin
+            return (origin, [expand_pattern(a) for a in args])
+
+        self.tp = tp
+        self.coercible = (
+            list(coercible) if coercible is not None else [(ty.Any, ty.Any)]
+        )
+        self.not_coercible = list(not_coercible) if not_coercible is not None else []
+        self.pattern = expand_pattern(tp)
+
+    def __call__(self, obj: ty.Any) -> T:
```

Comment:
Going by the docstring, it seems like this is actually:

```suggestion
    def __call__(self, obj: object) -> T | LazyField[T]:
```

1) You don't need `Any`. You're actually saying "all this needs to be is an object, and I'll either convert it or raise an exception". `object` is the correct type for that.
2) I think it's worth explicitly saying this could return a `Future` that emits the correct type (or that we'll convert when appropriate).

(We do still need to special-case `attr.NOTHING` for now, since you would have to use the private `attr._Nothing` type. `Literal[Nothing]` is broken: https://github.com/python-attrs/attrs/pull/983)

This seems like it will probably come up a lot. If so, we would want to represent this concept with a type alias like `Field[T]`:

```Python
Field: TypeAlias = T | LazyField[T]
```

Reply:
Ok, will look to tighten that up